// arco-cli/lib/teaLog.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const axios_1 = __importDefault(require("axios"));
const os_1 = __importDefault(require("os"));
const arco_cli_dev_utils_1 = require("arco-cli-dev-utils");
function teaLog(params) {
    var _a, _b;
    try {
        const globalInfo = (0, arco_cli_dev_utils_1.getGlobalInfo)();
        if (globalInfo && ((_a = globalInfo.host) === null || _a === void 0 ? void 0 : _a.arco)) {
            const message = {
                user: {
                    username: ((_b = globalInfo.userInfo) === null || _b === void 0 ? void 0 : _b.username) || '',
                },
                event: {
                    event: 'arco_cli',
                    time: +Date.now(),
                    params: Object.assign(Object.assign({}, params), { os: os_1.default.type(), node: process.version }),
                },
            };
            axios_1.default.post(`${globalInfo.host.arco}/material/api/log`, message).catch(() => { });
        }
    }
    catch (err) { }
}
exports.default = teaLog;
