// arco-cli-dev-utils/lib/MessageQueue.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const print_1 = __importDefault(require("./print"));
class MessageQueue {
    constructor({ description }) {
        this.description = description;
        this.info = [];
        this.success = [];
        this.warn = [];
        this.error = [];
    }
    /**
     * @param {'info' | 'success' | 'warn' | 'error'} type
     * @param  {string[]} messages
     */
    push(type, messages) {
        this[type] && this[type].push(messages);
    }
    flush() {
        const chalk = print_1.default.chalk;
        print_1.default.divider();
        (0, print_1.default)(`${this.description} ${chalk.green(`SUCCESS(${this.success.length})`)} ${chalk.yellow(`WARN(${this.warn.length})`)} ${chalk.red(`ERROR(${this.error.length})`)}`);
        this.info.forEach((messages) => print_1.default.info(...messages));
        this.success.forEach((messages) => print_1.default.success(...messages));
        this.warn.forEach((messages) => print_1.default.warn(...messages));
        this.error.forEach((messages) => print_1.default.error(...messages));
        print_1.default.divider();
    }
}
exports.default = MessageQueue;
